druid-shell/src/alert.rs

```diff
@@ -0,0 +1,451 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Alert dialogs.
+
+use std::borrow::Cow;
+
+/// A token that uniquely identifies an alert dialog.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
+pub struct AlertToken(usize);
+
+/// All the info needed to show and respond to an alert.
+#[derive(Debug, Clone)]
+pub(crate) struct AlertRequest {
+    pub(crate) token: AlertToken,
+    pub(crate) options: AlertOptions,
+}
+
+/// Contains the result of the alert after it was closed.
+#[derive(Debug, Clone)]
+pub struct AlertResponse {
+    token: AlertToken,
+    button: Option<AlertButton>,
```

Comment:
I presume that the button is used to resolve what action was taken?

My inclination would be to not use `AlertButton` for that, and instead have the user pass in some other token for each option? Although now that I write this out I see how annoying it might be.

I guess this comes down to my concern about using string equality to check for intent, I could imagine cases (maybe?) where a locale changes while a modal is up and the user checks against a changed string, or similar?

Perhaps a more reasonable concern is that this doesn't work as well when the buttons are localized, because you wan't have a `const` `AlertButton` sitting around.

So I'm not convinced either way, but would be curious to hear your thoughts.

Reply:
This is a pain point, I wrestled with various designs and I agree this is not perfect. Right now the idea is that you either have a `const AlertButton` or you store the localized `AlertButton` and check against that. This stored one wouldn't change when the locale changes, but it can be a bit of a pain to plumb around because it does not implement `Copy`.

I was thinking of also having some sort of `AlertButtonToken` but that runs into the same issue that commands/selectors do, in that this token would have to be a manually specified string to support `const`. Also having a token would make things less streamlined for const buttons.

Thinking about this some more now, perhaps there is a way to make this somewhat better in a way that wouldn't decrease the ergonomics of const buttons.

We could define that a const button's token is equal to its label. Then `AlertResult` could have `Option<AlertButtonToken>` and we could also support `AlertButton == AlertButtonToken` by using the button's underlying token for equality. To create localized buttons you would have to supply both a token to identify the button, and a label. The `AlertButtonToken` would implement `Copy` and be able to be `const`. The ergonomics win from having such an easily usable token might be greater than the ergonomics loss of having to define the tokens.

Reply:
Here's another consideration:

there are special cases, like 'ok' and 'cancel', where in general we want to let the platform provide the strings and handle the localization, (If possible?) and so we'd want to have some nice way of handling this.

some thoughts: we could maybe stick with the current mechanism, but have the "action identifier" or something be different from the display text? And maybe we could also have convenient `is_confirm()` and `is_cancel()` methods?

honestly... I defer to you here, you've been thinking about it more, I don't think that from a distance I'm going to come up with some sort of better solution.

Reply:
I'll play around a bit more with this API and see if anything seems more ergonomical. The idea of `is_cancel` is there already in form of `None` but I guess an additional method to `AlertResponse` wouldn't hurt much. A similar method for the primary button could also be worth it.

---

I don't think that letting the platform localize some buttons is a good idea. I've used apps like that and it's always a horrible user experience because the platform and app won't have a matching set of translated languages. Solving this is one of the key goals I personally have regarding localization.

If the app doesn't have the language but the platform has, then the alert message (and some buttons!) will be in English while OK/Cancel will be translated. The other way around you could have the message and some buttons translated and then suddenly some English OK/Cancel thrown in to the mix because the platform doesn't have that language. All of this gets even more insane once you mix RTL languages in.

The app should have full localization control. This will allow the user experience in the app to be consistent. If the app doesn't support a locale, then at least everything will be in a locale it supports. On the other hand if it does support a specific locale, then the platform won't be holding it back.

The user should also be able to override the language in the app. Nothing pisses me off more than when an app translates everything into my locale (at Google translate level no less) and has no way to change it. Also in today's globalized world many people have organization provided computers where the language is stuck at whatever the organization has chosen. For example my girlfriend works for a Danish organization and has a Danish language Windows - but she doesn't speak Danish. Not being able to override this choice in apps (or having some buttons suddenly be in Danish) is a very bad experience.

Reply:
That makes sense, the only annoyance is that we then want to ship localization files that have the same strings as the platform localization files for every supported locale, which isn't the end of the world.